### ISSUE TITLE
CA1508: false positive

### ISSUE THREAD

**Issue by emperador-ming** [dotnet/roslyn-analyzers#7152]
### Analyzer

**Diagnostic ID**: [CA1508](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1508): Avoid dead conditional code

### Analyzer source

**SDK**: [Built-in CA analyzers in .NET 5 SDK or later](https://learn.microsoft.com/dotnet/fundamentals/productivity/code-analysis)

**Version**: [SDK 8.0.101](https://dotnet.microsoft.com/download/dotnet/8.0)

```csharp
public class Record
{
    public DateTime? OurLastUpdateDate { get; set; }

    public DateOnly? TheirLastUpdateDate { get; set; }

    public bool HasPendingUpdates => OurLastUpdateDate switch
    {
        not null => DateOnly.FromDateTime(OurLastUpdateDate.Value) < TheirLastUpdateDate,
        null => true, // CA1508
    };
}
```
### Expected behavior
```csharp
public class Record
{
    public DateTime? OurLastUpdateDate { get; set; }

    public DateOnly? TheirLastUpdateDate { get; set; }

    public bool HasPendingUpdates => OurLastUpdateDate switch
    {
        not null => DateOnly.FromDateTime(OurLastUpdateDate.Value) < TheirLastUpdateDate,
        null => true, // No warning
    };
}
```
### As expected, these cases don't produce warnings
```csharp
public bool HasPendingUpdates => OurLastUpdateDate switch
{
    not null => DateOnly.FromDateTime(OurLastUpdateDate.Value) < TheirLastUpdateDate,
    _ => true, // No warning
};
```
```csharp
public bool HasPendingUpdates => OurLastUpdateDate switch
{
    not null => false,
    null => true, // No warning
};
```
```csharp
public bool HasPendingUpdates => OurLastUpdateDate switch
{
    not null =>TheirLastUpdateDate.HasValue,
    null => true, // No warning
};
```

**Comment by mavasani** [comment]
Seems like a special case to be handled in the analyzer to understand that the two branches of switch with `not null` and `null` checks produces identical branching code as branches with `not null` and `_` check, i.e. the `null` branch in the first case does not perform a redundant null check as the compiler optimizes that away.
